pkg/compute/store/boltdb/store.go

```diff
@@ -0,0 +1,439 @@
+func (s *Store) appendHistory(
+	tx *bolt.Tx,
+	updatedExecution store.Execution,
+	previousState store.ExecutionState, comment string) error {
+	historyBucket := tx.Bucket([]byte(BucketHistoryName))
+	executionHistoryBucket, err := historyBucket.CreateBucketIfNotExists([]byte(updatedExecution.ID))
+	if err != nil {
+		return err
+	}
+
+	historyEntry := store.ExecutionHistory{
+		ExecutionID:   updatedExecution.ID,
+		PreviousState: previousState,
+		NewState:      updatedExecution.State,
+		NewVersion:    updatedExecution.Version,
+		Comment:       comment,
+		Time:          updatedExecution.UpdateTime,
+	}
+	historyEntryData, err := json.Marshal(historyEntry)
+	if err != nil {
+		return err
+	}
+
+	// History entry keys within the execution bucket will be iterated over
+	// in lexographical order so we want to pad the numbers with leading 0s
+	// although this adds a hard limit of 999 history items for an execution
+	seqNum, err := executionHistoryBucket.NextSequence()
+	if err != nil {
+		return err
+	}
+	seq := fmt.Sprintf("%03d", seqNum)
+	return executionHistoryBucket.Put([]byte(seq), historyEntryData)
+}
+
+// DeleteExecution delete the execution, removes its history and removes it
+// from the job index (along with the job indexes bucket)
+func (s *Store) DeleteExecution(ctx context.Context, executionID string) error {
+	log.Ctx(ctx).Trace().
+		Str("ExecutionID", executionID).
+		Msg("boltdb.DeleteExecution")
+
+	return s.database.Update(func(tx *bolt.Tx) (err error) {
+		return s.deleteExecution(tx, executionID)
+	})
+}
+
+func (s *Store) deleteExecution(tx *bolt.Tx, executionID string) error {
+	var errorList error
+
+	execution, err := s.getExecution(tx, executionID)
+	if err != nil {
+		errorList = errors.Join(errorList, store.NewErrExecutionNotFound(executionID))
+	}
+
+	// Delete single execution entry
+	executionsBucket := tx.Bucket([]byte(BucketExecutionsName))
+	err = executionsBucket.Delete([]byte(executionID))
+	if err != nil {
+		errorList = errors.Join(errorList, err)
+	}
+
+	// Delete from job index
+	jobIndexBucket := tx.Bucket([]byte(BucketJobIndexName))
+	jobBucket := jobIndexBucket.Bucket([]byte(execution.Job.ID()))
+	if jobBucket != nil {
+		err = jobBucket.Delete([]byte(executionID))
+		if err != nil {
+			return err
+		}
+	}
+
+	// Delete the bucket with the execution-id within the history bucket
+	historyBucket := tx.Bucket([]byte(BucketHistoryName))
+	err = historyBucket.DeleteBucket([]byte(executionID))
+	if err != nil {
+		errorList = errors.Join(errorList, err)
+	}
+
+	return errorList
+}
+
+// Close ensures the database is closed cleanly
+func (s *Store) Close(ctx context.Context) error {
+	return s.database.Close()
+}
+
+func (s *Store) GetExecutionCount(ctx context.Context) (uint, error) {
+	log.Ctx(ctx).Trace().
+		Msg("boltdb.GetExecutionCount")
+
+	// TODO(ross) So, counting based on execution state .... seems like
+	// maintaining a prefix for a counter is excessive
```

Comment:
I just to make sure you saw my comment in your previous revision about calculating counts

Reply:
Yes, working on that today.